### DISCUSSION TITLE
Better git flow

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#365] [Ideas]
This is an unfinished proposal to improve how Dagger integrates with an existing git flow.

# Problem

*TLDR: Dagger must manage environment state correctly, while also integrating with a typical developer git. Achieving both requires tradeoffs. The current tradeoff can be improved.*

Dagger environments have state, which must be stored somewhere. Currently that somewhere is a shared directory at `$HOME/.dagger/store`. This shared directory is managed exclusively by Dagger.

This means that Dagger environments have their own lifecycle, distinct from the lifecycle of the application. For example, removing a git repository will not automatically remove the environments which depend on it; switching from multi-repo to mono-repo won’t affect how environments are organized; etc.

* The benefit of this decoupling is more flexibility: one environment can map to zero, one or more git repositories.
* The downside is more complexity for very common case: for example environments that map 1-1 to a git repository.

When a developer uses Dagger in the context of a single application repository (the way they use a Dockerfile or Docker Compose file), the decoupling forces them to do more work, because everything in their workflow is organized in the context of their git repository - except Dagger, because environments live outside the git repository.

Dagger does have *some* context-awareness:

    * `dagger new` defaults to the current directory name
    * `dagger —deployment` defaults to a lookup of deployments whose plans are the local directory
    * `dagger list` informs the user of which deployment is the “active deployment”, if any.

This is helpful, but the user flow still feels awkward:
* too much time is spent thinking about a) how all dagger deployments are organized, b) how the current repository is organized, and c) how to keep both in sync
* Context-awareness is limited to the current directory, and not to the underlying git repository, which would be more intuitive to developers

How to make Dagger integrate better in a typical git flow, without a) breaking state management, and b) breaking the experience for users which do not follow typical git flow?

# Solution

## Environments and workspaces

* Dagger runs its commands in isolated *environments*
* Environments are grouped together in a *workspace*
* Any git repository can be used as a workspace
* When using Dagger in the context of their git flow, a developer will use its current git repository as their Dagger workspace
* If no workspace is specified, `dagger` will automatically select the current git repo, enabling this pattern by default.
* If the current directory is not inside a git repo, a special namespace called `default` is used, enabling non-gitflow patterns

## Example

Using the default namespace

```
$ cd ~
$ dagger workspace
default
$ dagger list
example-react
example-s3
demo
throaway-test
$ dagger up demo
[...]
```

```
$ cd ~/src/acme-clothing/monorepo/frontend
$ dagger workspace
~/src/acme-clothing/monorepo
$ dagger list
dev
prod
staging
$ dagger up dev
[...]
```

## Architecture

Dagger state is split in two parts: global and contextual.

### Global state

The global state contains the persistent state of all environments across all local workspaces. Environments are addressable only by a globally unique ID.

The global state has 3 important properties:
1. It does something very basic (reliably save the internal state of each environment after each change);
2. with a very small API (CRUD on cue or json files);
3. its contents are location-independent, in particular: no references to local directories

These properties combined make it suitable for pluggable storage backends. By default the global state is stored in a local directory `~/.dagger/store`, but in the future it could be backed by remote storage services such as S3 or NFS.

### Contextual state

Contextual state is metadata “attached” to the current context git directory.

The exact implementation of *attachment* is not yet defined. Possible implementations include:
* Special file or directory at the root of the repository, requiring user to add it to `.gitignore`
* Special file or directory inside the `.git` directory (does not require `.gitignore`)
* Special `git config` values
* Special git head alongside user’s git branches.

Contextual state includes a named map of environments, each with a name, global ID, and input configuration.

This means:
  * `dagger list` may show entirely different results depending on context
  * `dagger -d FOO` may select entirely different deployments for the same value of FOO, depending on context
  * `dagger new` may select a different default name dependong on context (already true today I guess)
  * `dagger up` can only work from within the workspace which created the environment
